src/utils/webrtc/simplewebrtc/peer.js

```diff
@@ -372,22 +400,134 @@ Peer.prototype.end = function() {
 	this.pc.close()
 	this.handleStreamRemoved()
 	this.parent.off('localTrackReplaced', this.handleLocalTrackReplacedBound)
+	this.parent.off('localTrackEnabledChanged', this.handleLocalTrackEnabledChangedBound)
 }
 
 Peer.prototype.handleLocalTrackReplaced = function(newTrack, oldTrack, stream) {
+	this._pendingReplaceTracks.push({ newTrack, oldTrack, stream })
+
+	this._processPendingReplaceTracks()
+}
+
+/**
+ * Process pending replace track actions.
+ *
+ * All the pending replace track actions are executed from the oldest to the
+ * newest, waiting until the previous action was executed before executing the
+ * next one.
+ *
+ * The process may be stopped if the connection is lost, or if a track needs to
+ * be added rather than replaced, which requires a renegotiation. In both cases
+ * the process will start again once the connection is restablished.
+ *
+ * @returns {Promise} a Promise fulfilled when the processing ends; if it was
+ *          completed the resolved value is true, and if it was stopped before
+ *          finishing the resolved value is false.
+ */
+Peer.prototype._processPendingReplaceTracks = function() {
+	if (this._processPendingReplaceTracksPromise) {
+		return this._processPendingReplaceTracksPromise
+	}
+
+	this._processPendingReplaceTracksPromise = new Promise((resolve, reject) => {
+		for (let i = 0; i < this._pendingReplaceTracks.length; i++) {
+			if (this.pc.iceConnectionState === 'new') {
+				// Do not replace the tracks when the connection has not started
+				// yet, as Firefox can get "stuck" and not replace the tracks
+				// even if tried later again once connected.
+
+				resolve(false)
+
+				return
+			}
+
+			const pending = this._pendingReplaceTracks[i]
+
+			try {
+				this._replaceTrack(pending.newTrack, pending.oldTrack, pending.stream).then(() => {
+					this._pendingReplaceTracks.splice(i, 1)
+					i--
+				})
+			} catch (exception) {
+				this._pendingReplaceTracks.splice(i, 1)
+
+				// If the track is added instead of replaced a renegotiation
+				// will be needed, so stop replacing tracks.
+
+				resolve(false)
+
+				// A real try/catch rather than using Promise.catch is needed to
+				// be able to return from the catch.
+				return
+			}
+		}
+
+		resolve(true)
```

Comment:
so here the promise would resolve before we started cleaning up "pendingReplaceTracks" so a subsequent call might still have those in place.

how about using async/await on the `replaceTrack` call to make sure we wait until resolving ?

Reply:
Actually that is what was being done, `await this._replaceTrack`, which also should address your (very valid) concerns about the index above.

The problem is that [the linter complained that the promise was using an `async` executor](https://eslint.org/docs/rules/no-async-promise-executor). So I adjusted it... and I did it wrong ;-)

Reply:
I added back the `await` call and then extracted an `async` function and replaced the explicit `Promise` constructor with a call to that `async` function. That should fix the issue while keeping the linter happy.